### PR TITLE
Rework for v2

### PR DESCRIPTION
**Pull request by corny**
I'm currently the maintainer https://github.com/rylio/ytdl and I would like to deprecate that package and to improve this one.

* `Client` struct fetches video details, video streams, and deciphers URLs
* `Video` struct contains video and stream details
* `Downloader` struct offers methods to download videos into files
* Default HTTP client can be overridden for Proxy, TLS options, etc.
* Add Context to HTTP requests
* Never panic
* Use https://github.com/olekukonko/tablewriter for output formating

### CODE REVIEW

File: client.go [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+		return nil, err
+	}
+
+	resp, err = client.Do(req)
+	if err != nil {
+		return nil, err
+	}
+
+	if resp.StatusCode != http.StatusOK {
+		resp.Body.Close()
```
Comment (Julian-Chu) [inline client.go:96]: it seems to me that moving Body.Close() to  line:94 with defer would be better.

File: client.go [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+	}
+
+	eurl := "https://youtube.googleapis.com/v/" + id
+	resp, err := c.httpGet(ctx, "https://youtube.com/get_video_info?video_id="+id+"&eurl="+eurl)
+	if err != nil {
+		return nil, err
+	}
+
+	body, err := ioutil.ReadAll(resp.Body)
+	resp.Body.Close()
```
Comment (Julian-Chu) [inline client.go:40]: It seems to me that close should be  after error check with defer.

File: downloader/downloader.go [hunk 3]
```diff
@@ -0,0 +1,179 @@
...
+			return "", err
+		}
+		outputFile = filepath.Join(dl.OutputDir, outputFile)
+	}
+
+	return outputFile, nil
+}
+
+//Download : Starting download video by arguments.
+func (dl *Downloader) Download(ctx context.Context, outputFile string, v *youtube.Video, stream *youtube.Stream) error {
```
Comment (Julian-Chu) [inline downloader/downloader.go:42]: I recommend to change the signature to  (ctx, video, stream, outputFile), same in other download methods

**Comment by Julian-Chu** [comment]
I learn a lot  from this rework. IMHO it looks nice to move from draft to review.

File: downloader/downloader.go (hunk 3 above)
Comment (corny) [reply downloader/downloader.go]: Oh, you're right! done

File: client.go (hunk 2 above)
Comment (corny) [reply client.go]: done

**Comment by corny** [comment]
Currently there are methods with and without context:
* `GetVideo(url string)`
* `GetVideoContext(ctx context.Context, url string)`
* `Download(ctx context.Context, v *Video, stream *Stream)`

This is inconsistent. We should either only have methods with context, or have both methods with and without context. That means:
* `GetVideo(ctx context.Context, url string)`
* `Download(ctx context.Context, v *Video, stream *Stream)`
or 
* `GetVideo(url string)`
* `GetVideoContext(ctx context.Context, url string)`
* `Download(v *Video, stream *Stream)`
* `DownloadContext(ctx context.Context, v *Video, stream *Stream)`

File: client.go (hunk 1 above)
Comment (corny) [reply client.go:96]: Not in this case. The caller of this methods needs to read be body.

File: cmd/youtubedr/main.go [hunk 4]
```diff
@@ -0,0 +1,141 @@
...
+		log.Println("Skip server certificate verification")
+		httpTransport.TLSClientConfig = &tls.Config{
+			InsecureSkipVerify: true,
+		}
+	}
+
+	dl := ytdl.Downloader{
+		OutputDir: outputDir,
+	}
+	dl.Client.HTTPClient = &http.Client{Transport: httpTransport}
```
Comment (Julian-Chu) [inline cmd/youtubedr/main.go:104]: It seems to me that exposing Client struct may not be necessary.
Could we just use client as embedded struct  to let Downloader is the only one exported Struct
and provide constructor of  Downloader to initialize client and outputdir?

Comment (Julian-Chu) [reply cmd/youtubedr/main.go:104]: BTW, this comment is  the last one  I would like to discuss, then v2 could be merged and released

Comment (corny) [reply cmd/youtubedr/main.go]: Oh yes, that's a good idea.

File: downloader/downloader.go [hunk 5]
```diff
@@ -149,29 +149,18 @@ func (dl *Downloader) videoDLWorker(ctx context.Context, out *os.File, video *yo
...
-	_, err := io.Copy(ioutil.Discard, r)
-	if err != nil && err.Error() != youtube.ErrReadOnClosedResBody.Error() {
-		dl.logf("failed to exhaust reader: %s", err)
-	}
-	err = r.Close()
-	if err != nil {
-		dl.logf("response close err %s", err)
-	}
-}
-
```
Comment (corny) [inline downloader/downloader.go]: @Julian-Chu Do you agree that this method is not required at all?

Comment (Julian-Chu) [reply downloader/downloader.go]: yes, it's fine with me to remove this method

**Comment by Julian-Chu** [comment]
thanks for great job! I have no further comments.

**Pull request merged**